File: IsometricRPG/IsometricRPG/World/DungeonGenerator.swift
import CoreGraphics

/// A rectangular room placed within a chunk during dungeon generation.
struct Room {
    let x: Int      // local to chunk (0..<chunkSize)
    let y: Int
    let width: Int
    let height: Int

    var centerX: Int { x + width / 2 }
    var centerY: Int { y + height / 2 }
    var center: (Int, Int) { (centerX, centerY) }

    func intersects(_ other: Room, padding: Int = 1) -> Bool {
        return x - padding < other.x + other.width &&
               x + width + padding > other.x &&
               y - padding < other.y + other.height &&
               y + height + padding > other.y
    }
}

/// Generates dungeon layouts (rooms + corridors) for individual chunks.
/// Each chunk gets a deterministic layout based on its seed.
enum DungeonGenerator {

    /// Generate a set of rooms for a chunk. Returns rooms in local chunk coordinates.
    static func generateRooms(seed: UInt64, count: Int = Constants.roomsPerChunk) -> [Room] {
        var rng = SeededRNG(seed: seed &+ 0xDEADBEEF)
        var rooms: [Room] = []
        let size = Constants.chunkSize

        for _ in 0..<(count * 3) { // attempt up to 3x to place rooms
            guard rooms.count < count else { break }

            let w = Constants.roomMinSize + Int(rng.next() % UInt64(Constants.roomMaxSize - Constants.roomMinSize + 1))
            let h = Constants.roomMinSize + Int(rng.next() % UInt64(Constants.roomMaxSize - Constants.roomMinSize + 1))
            let rx = 1 + Int(rng.next() % UInt64(max(size - w - 2, 1)))
            let ry = 1 + Int(rng.next() % UInt64(max(size - h - 2, 1)))

            let candidate = Room(x: rx, y: ry, width: w, height: h)
            let overlaps = rooms.contains { $0.intersects(candidate, padding: 1) }
            if !overlaps {
                rooms.append(candidate)
            }
        }

        return rooms
    }

    /// Carve rooms and L-shaped corridors between them into a tile grid.
    /// The grid should already be filled with the base tile (e.g., dungeonWall).
    static func carve(rooms: [Room], into tiles: inout [[TileType]]) {
        // Carve rooms
        for room in rooms {
            for row in room.y..<(room.y + room.height) {
                for col in room.x..<(room.x + room.width) {
                    guard row >= 0, row < tiles.count, col >= 0, col < tiles[0].count else { continue }
                    tiles[row][col] = .dungeonFloor
                }
            }
        }

        // Connect rooms with corridors (each room to the next)
        for i in 0..<(rooms.count - 1) {
            let (ax, ay) = rooms[i].center
            let (bx, by) = rooms[i + 1].center
            carveCorridor(from: (ax, ay), to: (bx, by), into: &tiles)
        }

        // Place doorways at corridor-room transitions
        for room in rooms {
            placeDoorways(room: room, tiles: &tiles)
        }
    }

    private static func carveCorridor(from a: (Int, Int), to b: (Int, Int),
                                       into tiles: inout [[TileType]]) {
        let (ax, ay) = a
        let (bx, by) = b
        let w = Constants.corridorWidth

        // Horizontal leg
        let xRange = ax < bx ? ax...bx : bx...ax
        for col in xRange {
            for offset in 0..<w {
                let row = ay + offset - w / 2
                guard row >= 0, row < tiles.count, col >= 0, col < tiles[0].count else { continue }
                if tiles[row][col] == .dungeonWall {
                    tiles[row][col] = .corridor
                }
            }
        }

        // Vertical leg
        let yRange = ay < by ? ay...by : by...ay
        for row in yRange {
            for offset in 0..<w {
                let col = bx + offset - w / 2
                guard row >= 0, row < tiles.count, col >= 0, col < tiles[0].count else { continue }
                if tiles[row][col] == .dungeonWall {
                    tiles[row][col] = .corridor
                }
            }
        }
    }

    private static func placeDoorways(room: Room, tiles: inout [[TileType]]) {
        let edges = [
            (room.centerX, room.y - 1),          // top
            (room.centerX, room.y + room.height), // bottom
            (room.x - 1, room.centerY),           // left
            (room.x + room.width, room.centerY),  // right
        ]

        for (col, row) in edges {
            guard row >= 0, row < tiles.count, col >= 0, col < tiles[0].count else { continue }
            if tiles[row][col] == .corridor {
                tiles[row][col] = .doorway
            }
        }
    }
}

File: IsometricRPGMono/World/DungeonGenerator.cs
namespace IsometricRPG;

public struct Room
{
    public int X      { get; set; }
    public int Y      { get; set; }
    public int Width  { get; set; }
    public int Height { get; set; }

    public Room(int x, int y, int width, int height)
    {
        X = x; Y = y; Width = width; Height = height;
    }

    public int CenterX => X + Width  / 2;
    public int CenterY => Y + Height / 2;

    /// <param name="padding">Extra tile border around each room treated as occupied, preventing adjacent rooms from touching.</param>
    public bool Intersects(Room other, int padding = 1)
    {
        return X - padding < other.X + other.Width  &&
               X + Width  + padding > other.X        &&
               Y - padding < other.Y + other.Height  &&
               Y + Height + padding > other.Y;
    }
}

public static class DungeonGenerator
{
    /// <summary>
    /// Generates a dungeon layout into a TileType[,] grid of size (size x size).
    /// Grid is indexed as [row, col].
    /// </summary>
    public static List<Room> Generate(TileType[,] tiles, int size, SeededRNG rng)
    {
        int rows = tiles.GetLength(0);
        int cols = tiles.GetLength(1);

        // Fill with walls
        for (int r = 0; r < rows; r++)
            for (int c = 0; c < cols; c++)
                tiles[r, c] = TileType.Wall;

        var rooms = new List<Room>();

        for (int attempt = 0; attempt < Constants.RoomsPerChunk * 10; attempt++)
        {
            if (rooms.Count >= Constants.RoomsPerChunk) break;

            int w = Constants.RoomMinSize + (int)(rng.Next() % (ulong)(Constants.RoomMaxSize - Constants.RoomMinSize + 1));
            int h = Constants.RoomMinSize + (int)(rng.Next() % (ulong)(Constants.RoomMaxSize - Constants.RoomMinSize + 1));
            int x = 1 + (int)(rng.Next() % (ulong)(cols - w - 2));
            int y = 1 + (int)(rng.Next() % (ulong)(rows - h - 2));

            var candidate = new Room(x, y, w, h);

            bool overlaps = rooms.Any(r => r.Intersects(candidate));
            if (overlaps) continue;

            CarveRoom(tiles, candidate);
            rooms.Add(candidate);

            if (rooms.Count > 1)
            {
                var prev = rooms[rooms.Count - 2];
                ConnectRooms(tiles, prev, candidate, rng);
            }
        }

        return rooms;
    }

    private static void CarveRoom(TileType[,] tiles, Room room)
    {
        for (int r = room.Y; r < room.Y + room.Height; r++)
            for (int c = room.X; c < room.X + room.Width; c++)
                if (r >= 0 && r < tiles.GetLength(0) && c >= 0 && c < tiles.GetLength(1))
                    tiles[r, c] = TileType.Floor;
    }

    private static void ConnectRooms(TileType[,] tiles, Room a, Room b, SeededRNG rng)
    {
        // Randomly choose whether to go horizontal-first or vertical-first
        bool hFirst = (rng.Next() % 2UL) == 0UL;

        if (hFirst)
        {
            CarveCorridor(tiles, a.CenterX, a.CenterY, b.CenterX, a.CenterY, horizontal: true);
            CarveCorridor(tiles, b.CenterX, a.CenterY, b.CenterX, b.CenterY, horizontal: false);
        }
        else
        {
            CarveCorridor(tiles, a.CenterX, a.CenterY, a.CenterX, b.CenterY, horizontal: false);
            CarveCorridor(tiles, a.CenterX, b.CenterY, b.CenterX, b.CenterY, horizontal: true);
        }
    }

    private static void CarveCorridor(TileType[,] tiles, int x1, int y1, int x2, int y2, bool horizontal)
    {
        int rows = tiles.GetLength(0);
        int cols = tiles.GetLength(1);
        int half = Constants.CorridorWidth / 2;

        if (horizontal)
        {
            int minX = Math.Min(x1, x2);
            int maxX = Math.Max(x1, x2);
            for (int c = minX; c <= maxX; c++)
                for (int offset = -half; offset <= half; offset++)
                {
                    int r = y1 + offset;
                    if (r >= 0 && r < rows && c >= 0 && c < cols)
                        tiles[r, c] = TileType.Floor;
                }
        }
        else
        {
            int minY = Math.Min(y1, y2);
            int maxY = Math.Max(y1, y2);
            for (int r = minY; r <= maxY; r++)
                for (int offset = -half; offset <= half; offset++)
                {
                    int c = x1 + offset;
                    if (r >= 0 && r < rows && c >= 0 && c < cols)
                        tiles[r, c] = TileType.Floor;
                }
        }
    }
}
